### PR TITLE
make amp embed cleaner embed facebook videos from non guardian sources

### PR DESCRIPTION
**Pull request by michaelwmcnamara**
This is a change to the amp embed cleaner so that it will handle embedded facebook videos that are not from the guardian. For example [this page](https://amp.theguardian.com/global/commentisfree/2017/feb/16/first-dates-going-dutch-bill-splitting-channel-4#development=1) contains a facebook video embed from channel4.

This change fixes a [trello card](https://trello.com/c/pumy9Trr/228-non-guardian-amp-facebook-video-embeds-are-causing-validation-to-fail)

This change only effects amp pages
all tests pass 
validate-amp tests all pass

## Tested in CODE?
No.

cc @guardian/dotcom-platform

### CODE REVIEW

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala [hunk 1]
```diff
@@ -50,24 +50,25 @@ case class AmpEmbedCleaner(article: Article) extends HtmlCleaner {
...
-  case class FacebookExternalVideo(override val videoId: String) extends AmpExternalVideo(videoId, "amp-facebook", facebookAttributes(videoId))
+  case class FacebookExternalVideo(organisation: String, override val videoId: String) extends AmpExternalVideo(videoId, "amp-facebook", facebookAttributes(organisation, videoId))
 
 
   object AmpExternalVideo {
     def getAmpExternalVideoByUrl(videoUrl: String): Option[AmpExternalVideo] = {
       val youtubePattern = """^https?:\/\/www\.youtube\.com\/watch\?v=([^#&?]+).*""".r
       val vimeoPattern = """^https?:\/\/vimeo\.com\/(\d+).*""".r
-      val facebookPattern = """^https?:\/\/www\.facebook\.com\/theguardian\/videos\/(\d+)\/""".r
+      //val facebookPattern = """^https?:\/\/www\.facebook\.com\/theguardian\/videos\/(\d+)\/""".r
```
Comment (gidsg) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: remove commented out line.

File: common/test/views/support/cleaner/AmpEmbedCleanerTest.scala [hunk 2]
```diff
@@ -193,9 +193,22 @@ class AmpEmbedCleanerTest extends FlatSpec with Matchers {
...
+    result.getElementsByTag("amp-facebook").attr("width") should be("5")
+    result.getElementsByTag("amp-facebook").attr("height") should be("3")
+    result.getElementsByTag("amp-facebook").attr("layout") should be("responsive")
+  }
+
+
   "AmpEmbedCleaner" should "not replace an iframe in a fake Facebook video-element with an amp-facebook element" in {
     val result = cleanDocumentWithVideos(
-      "https://www.facebook.com/nottheguardian/videos/123456",
       "https://facbook.com.zz/theguardian/123456/",
```
Comment (gidsg) [inline common/test/views/support/cleaner/AmpEmbedCleanerTest.scala]: I don't think you really have to so extensively test your regex! But I suppose it does not do any harm

**Review by gidsg — CHANGES_REQUESTED** [review]

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala (hunk 1 above)
Comment (michaelwmcnamara) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: oops! Thanks, good spot!

File: common/test/views/support/cleaner/AmpEmbedCleanerTest.scala (hunk 2 above)
Comment (michaelwmcnamara) [inline common/test/views/support/cleaner/AmpEmbedCleanerTest.scala]: Thats actually a good point! I'll strip out the extra pages as they dont really add anything.

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala [hunk 3]
```diff
@@ -50,24 +50,24 @@ case class AmpEmbedCleaner(article: Article) extends HtmlCleaner {
...
-  case class FacebookExternalVideo(override val videoId: String) extends AmpExternalVideo(videoId, "amp-facebook", facebookAttributes(videoId))
+  case class FacebookExternalVideo(organisation: String, override val videoId: String) extends AmpExternalVideo(videoId, "amp-facebook", facebookAttributes(organisation, videoId))
 
 
   object AmpExternalVideo {
     def getAmpExternalVideoByUrl(videoUrl: String): Option[AmpExternalVideo] = {
       val youtubePattern = """^https?:\/\/www\.youtube\.com\/watch\?v=([^#&?]+).*""".r
       val vimeoPattern = """^https?:\/\/vimeo\.com\/(\d+).*""".r
-      val facebookPattern = """^https?:\/\/www\.facebook\.com\/theguardian\/videos\/(\d+)\/""".r
+      val facebookPattern = """^https?:\/\/www\.facebook\.com\/(\w+)\/videos\/(\d+)\/""".r
```
Comment (SiAdcock) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: Is there any chance the organisation might contain non-alphabetic characters such as digits or a `-`?

Comment (michaelwmcnamara) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: w+ matches digits and letters.
As far as I my powers of google can tell me - facebook usernames (which is where that bit comes from) only allow alphanumeric chars. You can include full stops, but they won't be recognized as part of the username so likely they are not included in the url either.
https://www.facebook.com/help/105399436216001

File: common/test/views/support/cleaner/AmpEmbedCleanerTest.scala (hunk 2 above)
Comment (michaelwmcnamara) [inline common/test/views/support/cleaner/AmpEmbedCleanerTest.scala]: @gidsg  These change requests should be resolved in the latest commit.
Cheers

Mike

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala (hunk 3 above)
Comment (gidsg) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: @michaelmcnamara I think you need a wider regex to include hyphens. I think you could use `((\w|\W))` I tested this on http://rubular.com/

Comment (gidsg) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: @michaelmcnamara you should allow for `.` in your regex e.g. https://www.facebook.com/gideon.goldberg

Comment (michaelwmcnamara) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: @gidsg @SiAdcock  Just made a change to cater for usernames with '.' (good spot both of you!).
Let me know if this is ok and I'll merge it tomorrow morning.

Comment (gidsg) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: @michaelmcnamara did you commit this, I don't see any new commit?

File: common/app/views/support/cleaner/AmpEmbedCleaner.scala [hunk 4]
```diff
@@ -63,7 +63,7 @@ case class AmpEmbedCleaner(article: Article) extends HtmlCleaner {
     def getAmpExternalVideoByUrl(videoUrl: String): Option[AmpExternalVideo] = {
       val youtubePattern = """^https?:\/\/www\.youtube\.com\/watch\?v=([^#&?]+).*""".r
       val vimeoPattern = """^https?:\/\/vimeo\.com\/(\d+).*""".r
-      val facebookPattern = """^https?:\/\/www\.facebook\.com\/(\w+)\/videos\/(\d+)\/""".r
+      val facebookPattern = """^https?:\/\/www\.facebook\.com\/([\w.-]+)\/videos\/(\d+)\/""".r
```
Comment (gidsg) [inline common/app/views/support/cleaner/AmpEmbedCleaner.scala]: I don't think `-` is allowed but `.` is. This is according to the guidelines you linked to.

**Comment by michaelwmcnamara** [comment]
@gidsg  Hi, this should now only allow chars, numbers or '.' chars.

File: common/test/views/support/cleaner/AmpEmbedCleanerTest.scala [hunk 5]
```diff
@@ -193,12 +193,35 @@ class AmpEmbedCleanerTest extends FlatSpec with Matchers {
...
+    result.getElementsByTag("amp-facebook").attr("layout") should be("responsive")
+  }
+
+  "AmpEmbedCleaner" should "replace a facebook non-guardian video embed, containing a '.' char in the username, with a valid amp-facebook video embed" in {
+    val facebookOrganisationId = "Channel.4"
+    val faceookVideoId = "10154084521542330"
+    val facebookVideoUrl = s"https://www.facebook.com/$facebookOrganisationId/videos/$faceookVideoId/"
+    val result = cleanDocumentWithVideos(facebookVideoUrl)
+    result.getElementsByTag("amp-facebook").size should be(1)
+    result.getElementsByTag("amp-facebook").attr("data-href") should be(s"https://www.facebook.com/$facebookOrganisationId/videos/$faceookVideoId")
```
Comment (gidsg) [inline common/test/views/support/cleaner/AmpEmbedCleanerTest.scala]: I would not bother repeating all these other tests for this additional case. Just the first `result.getElementsByTag("amp-facebook").size should be(1)` I think is sufficient.

Comment (michaelwmcnamara) [inline common/test/views/support/cleaner/AmpEmbedCleanerTest.scala]: good point.

**Comment by michaelwmcnamara** [comment]
@gidsg  Hi mate, thats been tidied up now.

**Review by gidsg — APPROVED** [review]

**Comment by gidsg** [comment]
Looks good to me once you rebase to get tests to pass. 👍

**Comment by michaelwmcnamara** [comment]
@gidsg  Thanks for your help with this! Shipping now. :-)

**Pull request merged**